tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -34,6 +41,352 @@
+   * buffer into the tensor, starting at the buffer's current position.
+   * The position of the buffer is then incremented by <i>n</i>.
+   *
+   * @param dataType the tensor datatype.
+   * @param shape the tensor shape.
+   * @param data a buffer containing the tensor data.
+   *
+   * @throws  IllegalArgumentException
+   *          If the tensor datatype or shape is not compatible with the buffer
+   */
+  public static Tensor create(DataType dataType, long[] shape, ByteBuffer data) {
+    switch(dataType) {
+      case FLOAT:
+      case DOUBLE:
+      case INT32:
+      case INT64:
+      case BOOL:
+        int elemSize = elemByteSize(dataType);
+        if(data.remaining() / elemSize != numElements(shape)) {
+          throw incompatibleBuffer(shape);
+        }
+        break;
+      case STRING:
+      default:
+        // not all types are checked
+        break;
+    }
+
+    Tensor t = create(dataType, shape, data.remaining());
+    try {
+      ByteBuffer dst = t.buffer();
+      dst.put(data);
+      return t;
+    }
+    catch(RuntimeException e) {
+      delete(t.nativeHandle);
+      throw e;
+    }
+  }
+
+  private static Tensor create(DataType dataType, long[] shape, long byteSize) {
+    Tensor t = new Tensor();
+    t.dtype = dataType;
+    t.shapeCopy = Arrays.copyOf(shape, shape.length);
+    t.nativeHandle = allocate(t.dtype.c(), t.shapeCopy, byteSize);
+    return t;
+  }
+
+  /**
+   * Gets the size (in bytes) of the tensor data.
+   *
+   * <p>Use this size information when constructing a byte buffer for {@code #writeTo(...)}.
+   */
+  public int byteSize() {
+    return buffer().remaining();
+  }
+
+  /**
+   * Gets the total number of elements in the tensor.
+   *
+   * <p>Use this size information when constructing a typed buffer for {@code #writeTo(...)}.
+   *
+   * @return the number of elements in a flattened (1-D) view of the tensor.
+   */
+  public int numElements() {
+    switch(dataType()) {
+      case FLOAT:
+      case DOUBLE:
+      case INT32:
+      case INT64:
+      case BOOL:
+        return byteSize() / elemByteSize(dataType());
+      case STRING:
+      default:
+        throw new UnsupportedOperationException("unsupported operation on a tensor of DataType "
+                + dataType());
+    }
+  }
+
+  /**
+   * Write the tensor data into the given buffer.
+   *
+   * <p>This method copies {@code numElements()} elements to the buffer.
+   *
+   * <p>This method may be used to read tensor data of type {@link DataType#INT32}.
+   *
+   * @param dst the destination buffer
+   *
+   * @throws  BufferOverflowException
+   *          If there is insufficient space in the given buffer
+   *          for the data in this tensor
+   * @throws  IllegalArgumentException
+   *          If the tensor datatype is not {@link DataType#INT32}
+   */
+  public void writeTo(IntBuffer dst) {
+    ByteBuffer src = buffer();
+    if(dtype != DataType.INT32) {
```

Comment:
In all the `writeTo` methods, maybe check consistency before calling `buffer()`?